### PR TITLE
fix(user-tenant-service): secure PATCH and PUT users API

### LINKED ISSUE #2004: Privilege escalation issue in PATCH and PUT users api
**Describe the bug**

As per privilege escalation rules, no user should be able to change his own role or its details on his own. This must be restricted. However its currently not checking for this. This is a security issue in that API.

**To Reproduce**

Check this here-https://github.com/sourcefuse/loopback4-microservice-catalog/blob/93a7f917f1ea5790547c939d0a9f56b6f725745b/services/user-tenant-service/src/services/user-operations.service.ts#L138

**Expected behavior**

1. A user should not be able to change his own role.
2. A user with permissions to edit role, should be able to update only those permissions which he himself has. He should not be able to give or takeaway any permission that he himself doesn't have.
3. A user should not be able to delete any role which has permissions higher than himself even though he has delete role permission.
4. A user should not be able to delete his own role.

### PR DESCRIPTION
**Pull request by Tyagi-Sunny**
## Description
secure PATCH and PUT users API

Fixes #2004 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] Performed a self-review of my own code
- [ ] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/user-tenant-service/src/controllers/tenant-role.controller.ts [hunk 1]
```diff
@@ -111,6 +117,13 @@ export class TenantRoleController {
       throw new HttpErrors.Forbidden(' Role with same name already exist');
     }
     role.tenantId = id ?? '';
+
+    const isAllowed =
```
Comment (samarpan-b) [inline services/user-tenant-service/src/controllers/tenant-role.controller.ts:121]: Can we move this to a service class please ? Also we need to make sure that its overridable logic.

File: services/user-tenant-service/src/controllers/tenant-role.controller.ts [hunk 2]
```diff
@@ -168,6 +187,12 @@ export class TenantRoleController {
     @param.path.string('roleId') roleId: string,
     @param.query.object('where', getWhereSchemaFor(Role)) where?: Where<Role>,
   ): Promise<void> {
+    const userTenant = await this.userTenantRepository.findById(
+      this.currentUser.userTenantId,
+    );
+    if (userTenant.roleId === roleId) {
+      throw new HttpErrors.Forbidden('User Not Allowed to delete its own Role');
+    }
     return this.roleRepository.deleteById(roleId);
```
Comment (samarpan-b) [inline services/user-tenant-service/src/controllers/tenant-role.controller.ts:196]: A user should not be allowed to delete a role which contains permissions which he himself doesnt have.

File: services/user-tenant-service/src/services/user-operations.service.ts [hunk 3]
```diff
@@ -176,6 +176,9 @@ export class UserOperationsService {
     currentUser: IAuthUserWithPermissions,
   ) {
     const utData: Partial<UserTenant> = {};
+    if (id === currentUser.id) {
```
Comment (samarpan-b) [inline services/user-tenant-service/src/services/user-operations.service.ts:179]: why a blanket ban here ? A user should be able to edit his other profile details.

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/user-tenant-service/src/controllers/tenant-role.controller.ts (hunk 2 above)
Comment (Tyagi-Sunny) [reply services/user-tenant-service/src/controllers/tenant-role.controller.ts:196]: ok , noted

File: services/user-tenant-service/src/controllers/tenant-role.controller.ts (hunk 1 above)
Comment (Tyagi-Sunny) [reply services/user-tenant-service/src/controllers/tenant-role.controller.ts:121]: this only include crud operation directly from repository, do we create a separate service for it?

File: services/user-tenant-service/src/controllers/tenant-role.controller.ts (hunk 2 above)
Comment (Tyagi-Sunny) [reply services/user-tenant-service/src/controllers/tenant-role.controller.ts:196]: done

File: services/user-tenant-service/src/controllers/tenant-role.controller.ts (hunk 1 above)
Comment (samarpan-b) [reply services/user-tenant-service/src/controllers/tenant-role.controller.ts:121]: create a tenant-operations service.

Comment (Tyagi-Sunny) [reply services/user-tenant-service/src/controllers/tenant-role.controller.ts:121]: ok sir

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: services/user-tenant-service/src/services/user-operations.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (samarpan-b) [reply services/user-tenant-service/src/services/user-operations.service.ts:179]: Not allowed @akshatdubeysf . Another super admin will have to do that. Thats how a system will always have a super admin.

Comment (Tyagi-Sunny) [reply services/user-tenant-service/src/services/user-operations.service.ts]: updated sir

[the remark this answers is not in the document]
Comment (akshatdubeysf) [reply services/user-tenant-service/src/services/user-operations.service.ts]: If we take example of something like GitHub or Slack, isn't that something that should be allowed?

[the remark this answers is not in the document]
Comment (akshatdubeysf) [reply services/user-tenant-service/src/services/user-operations.service.ts]: Considering we are developing this for the community

[the remark this answers is not in the document]
Comment (akshatdubeysf) [reply services/user-tenant-service/src/services/user-operations.service.ts]: @samarpan-b

[the remark this answers is not in the document]
Comment (akshatdubeysf) [reply services/user-tenant-service/src/services/user-operations.service.ts]: Ok

File: services/user-tenant-service/src/services/tenant-operations.service.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      this.currentUser.userTenantId,
+    );
+    if (userTenant.roleId === roleId || !isAllowed) {
+      throw new HttpErrors.Forbidden('Not Allowed');
+    }
+    return this.roleRepository.deleteById(roleId);
+  }
+  // it check whether user has all the permission that he/she is trying to Edit/Create/Delete
+  private _isAllowed(role: Partial<Role>): Boolean {
+    const isAllowed =
```
Comment (samarpan-b) [inline services/user-tenant-service/src/services/tenant-operations.service.ts:63]: We also need to check if permission is true or not.

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/user-tenant-service/src/services/tenant-operations.service.ts (hunk 4 above)
Comment (samarpan-b) [reply services/user-tenant-service/src/services/tenant-operations.service.ts:63]: add a test case for this too.

Comment (Tyagi-Sunny) [reply services/user-tenant-service/src/services/tenant-operations.service.ts:63]: test case already added

Comment (Tyagi-Sunny) [reply services/user-tenant-service/src/services/tenant-operations.service.ts:63]: what's meaning of permission true

Comment (samarpanB) [reply services/user-tenant-service/src/services/tenant-operations.service.ts:63]: Ah this is from token. Then not needed.

**Review by samarpanB — APPROVED** [review]

**Review by samarpan-b — APPROVED** [review]

**Review by akshatdubeysf — APPROVED** [review]

**Pull request merged**